Implement `Automationinterface` class

ahuca opened the pull request.

ahuca committed:
chore: added `TwinGet.AutomationInterface` project

ahuca committed:
feat: added `AutomationInterface`

ahuca committed:
chore: restructured project folder

ahuca committed:
feat: implemented the initial form of `AutomationInterface`

- Including constructor, destructor and dispose pattern
 - Added a simple unit test

ahuca committed:
chore: reorder content of `SupportedProgIds` List

Putting TcXaeShell on top of the list so that `AutomationInterface` prioritize using TwinCAT XAE (over Visual Studio) when creating a DTE instance

ahuca committed:
ci: added dotnet restore to build script

ahuca committed:
Merge remote-tracking branch 'origin/main' into 1-implement-automationinterface-class

ahuca committed:
refactor: fixed line ending for `AutomationInterface.cs`

ahuca committed:
refactor: improved TwinCAT integration check

Moved message filter revocation from `Dispose()` to `CleanUp()`

ahuca committed:
refactor: Added `DisposeOnCancelEvent` to `AutomationInterface`

ahuca committed:
Merge remote-tracking branch 'origin/main' into 1-implement-automationinterface-class

ahuca committed:
ci: Fixed `CI.yml` invalid syntax for actions/checkout and added nuget.config

ahuca committed:
Merge remote-tracking branch 'origin/main' into 1-implement-automationinterface-class

ahuca committed:
test: Added more tests to AutomationInterfaceTests

ahuca committed:
feat: AutomationInterface throws custom exception when failing to create a DTE instance

ahuca committed:
docs: Added documentation on self-hosted runner setup

ahuca committed:
chore: Added .editorconfig

[Based on dotnet runtime team's coding style](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md)

ahuca committed:
refactor: Made `CouldNotCreateTwinCatDte` exception public

ahuca committed:
refactor: Ignore code analysis in `MessageFilter`

ahuca committed:
test: Removed dummy test

ahuca committed:
refactor: Improved `TryInitializeDte`

Removed method `DisposeOnCancelEvent`, which was added for testing purpose.

Colho commented on src/TwinGet.Cli/TwinGet.Cli.csproj:
```diff
@@ -6,9 +6,10 @@
     <Description>TwinGet Command Line Interface</Description>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net7.0</TargetFramework>
+    <TargetFramework>net7.0-windows</TargetFramework>
```
Is the windows framework enforced by something?

ahuca replied:
Beckhoff's Automation Interface can only be supported by Windows, so this whole project is limited to Windows platform.

Colho commented on src/TwinGet.AutomationInterface/AutomationInterface.cs:
```diff
@@ -0,0 +1,100 @@
+                }
+                catch { continue; }
+
+                if (dte is null)
```
to simplify continue usage, could move this if directly under the dte creation and inline it?
if (dte is null) { continue; }

Colho commented on src/TwinGet.AutomationInterface/AutomationInterface.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            MessageFilter.Register();
+            _dte = TryInitializeDte(out _progId);
+            if (_dte is null) { throw new CouldNotCreateTwinCatDte("Is TwinCAT installed in this system?"); }
```
Depending on how this message is used, would be better to give more context, that the DTE creation failed, and as an additive tip to make sure TwinCAT is installed.

Colho reviewed (approved):
Can be merged, minor nitpicks and questions

ahuca committed:
refactor: Added more information and tip when throwing `CouldNotCreateTwinCatDte` exception

Also addressed PR comment https://github.com/ahuca/TwinGet/pull/2#discussion_r1416202660
